File: assignme.js
var request = require('request');
var fs = require("fs");
var enterprise_url = "https://github.ncsu.edu/api/v3/";
// var enterprise_url = 'https://api.github.com/';

// reading the token from file - Synchronous read
var token = fs.readFileSync('token');
var token = token.toString();

//**************************************************************
//**************************************************************
//*************** Creation / Assignment functions **************
//**************************************************************
//**************************************************************

/**
 * Creates repository under the given org and given name
 */
function createRepo(name, org) {
  var options = {
    // url: ' https://api.github.com/user/repos',
    url: enterprise_url + 'orgs/' + org + '/repos',
    method: 'POST',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    },
    json: {
      "name": name,
      "auto_init": true,
      "public": true,
      "gitignore_template": "nanoc"
    }
  };

  request(options, function(error, response, body) {
    console.log("Created repository: " + name);
  });
}

/**
 * Deletes a repository under the given org
 * Notice: if using OAuth, the `delete_repo` scope is required
 *
 * @param repoName repository name
 * @param org Organization name
 */
function deleteRepo(repoName, org) {
  var options = {
    url: enterprise_url + 'repos/' + org + '/' + repoName,
    method: 'DELETE',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    }
  };

  request(options, function(error, response, body) {
    console.log("Deleted repository: " + repoName);
  });
}

/**
 * Created a team under the given org
 * @param teamName Team name
 * @param org Organization name
 */
function createTeam(teamName, org) {
  var options = {
    url: enterprise_url + 'orgs/' + org + '/teams',
    method: 'POST',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    },
    json: {
      "name": teamName,
      "auto_init": true,
      "public": true,
      "gitignore_template": "nanoc"
    }
  };

  request(options, function(error, response, body) {
    var ID = body.id;
    return ID;
  });
}


/**
 * Adds the given repository to the given team
 *
 * @param repoName Repository name
 * @param teamName Team name
 * @param org Organization name
 */
function addRepoToTeam(repoName, teamName, org) {
  var options = {
    url: enterprise_url + 'teams/' + getTeamId(teamName, org) + '/repos/' + org + '/' + repoName,
    method: 'PUT',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    }
  };

  request(options, function(error, response, body) {
    console.log("Added repository: " + repoName + " to Team: " + teamName + " in " + org + "organization");
  });
}

/**
 * Deletes a team, given it's id
 *
 * @param id Team id
 */
function deleteTeam(id) {
  var options = {
    url: enterprise_url + 'teams/' + id,
    method: 'DELETE',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    }
  };

  request(options, function(error, response, body) {
    console.log("Deleted team: " + id.toString());
    console.log(error);
  });
}


/**
 * Gets team data for all teams under the given organization
 *
 * @param org Organization name
 */
var pageNumber = 1;

function getTeamData(org) {
  var options = {
    url: enterprise_url + 'orgs/' + org + '/teams?page=' + pageNumber + '&per_page=100',
    method: 'GET',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    }
  };

  request(options, function(error, response, body) {
    fs.appendFile("team.json", body, function(err) {});
  });
  if (pageNumber > 10)
    return;
  setTimeout(function() {
    console.log('Blah blah blah blah extra-blah');
    pageNumber++;
    getTeamData(org);
  }, 5000);
}

//Makes correction to json files after merging them
function writeJSON() {
  var previousData = fs.readFileSync('./team.json', 'utf8').toString();
  previousData = previousData.split('[]').join("");
  previousData = previousData.split('][').join(",");
  fs.writeFile('team.json', previousData, function() {});
}


/**
 * Parses the team data to get team id for the given team name
 *
 * @param teamName Team name
 * @param org Organization name
 */
function getTeamId(teamName, org) {
  var obj = JSON.parse(fs.readFileSync('./team.json', 'utf8'));
  for (var i = 0; i < obj.length; i++) {
    if (obj[i].name == teamName)
      return obj[i].id.toString();
  }
}


/**
 * Removes a repository from a team inside the given org
 *
 * @param repoName Repository name
 * @param teamName Team name
 * @param org Organization name
 */
function removeTeamRepository(repoName, teamName, org) {
  var options = {
    url: enterprise_url + 'teams/' + getTeamId(teamName, org) + '/repos/' + org + '/' + repoName,
    method: 'DELETE',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    }
  };

  request(options, function(error, response, body) {
    console.log("Deleted team repository: " + teamName);
  });
}


/**
 * Adds a user to the given team
 *
 * @param username username of the user
 * @param teamName Team name
 * @param org Organization name
 */
function addUserToTeam(username, teamName, org) {
  var options = {
    url: enterprise_url + 'teams/' + getTeamId(teamName, org) + '/memberships/' + username,
    method: 'PUT',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    }
  };

  request(options, function(error, response, body) {
		// console.log(getTeamId(teamName, org));
    console.log("Added " + username + " to team: " + teamName);
  });
}


/**
 * Removes a user from the given team.
 *
 * @param username The user's username
 * @param teamName Team name
 * @param org Organization name
 */
function removeUserFromTeam(username, teamName, org) {
  var options = {
    url: enterprise_url + 'teams/' + getTeamId(teamName, org) + '/memberships/' + username,
    method: 'DELETE',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    }
  };

  request(options, function(error, response, body) {
    console.log("Removed " + username + " from team: " + teamName);
  });
}

/**
 * Updates team repository permissions
 *
 * @param repoName repository name
 * @param teamName team name
 * @org orgnaization name
 * @permission new permission level : "pull", "push", "admin"
 */
function updateTeamRepositoryPermissions(repoName, teamName, org, permission) {
  var options = {
    url: enterprise_url + 'teams/' + getTeamId(teamName, org) + '/repos/' + org + '/' + repoName,
    method: 'PUT',
    headers: {
      "User-Agent": "AssignMe",
      "content-type": "application/json",
      "Authorization": token
    },
    json: {
      "permission": permission
    }
  };

  request(options, function(error, response, body) {
		// console.log(error);
    console.log("Updated permission on repository: " + repoName + " to " + permission);
  });
}
// updateTeamRepositoryPermissions("3", "3", "assignme", "push");

//readFILE line by line and assign the teams to ids.
function assignAll(inputFile, organization) {
  var fs = require('fs'),
    readline = require('readline'),
    instream = fs.createReadStream(inputFile),
    outstream = new(require('stream'))(),
    rl = readline.createInterface(instream, outstream);

  rl.on('line', function(line) {
    // console.log("username: " + line.split(":")[0]);
    // console.log("reponame: " + line.split(":")[1]);
    var username = line.split(":")[0];
    var teamname = line.split(":")[1];
    addUserToTeam(username, teamname, organization);
  });

  rl.on('close', function(line) {
    // console.log('done reading file.');
  });
}

//readFILE line by line and change access of each user to "permission" parameter
function changeAccess(inputFile, organization, permission) {
  var fs = require('fs'),
    readline = require('readline'),
    instream = fs.createReadStream(inputFile),
    outstream = new(require('stream'))(),
    rl = readline.createInterface(instream, outstream);

  rl.on('line', function(line) {
    // console.log("username: " + line.split(":")[0]);
    // console.log("reponame: " + line.split(":")[1]);
    var username = line.split(":")[0];
    var teamname = line.split(":")[1];
    updateTeamRepositoryPermissions(teamname, teamname, organization, permission);
  });

  rl.on('close', function(line) {
    // console.log('done reading file.');
  });
}



//**************************************************************
//**************************************************************
//************************** Testing****************************
//**************************************************************
//**************************************************************

// var usernames = ["theferrit32", "shorsher", "chbrown13", "sheckman", "jctetter", "ssmirr"];

// create team and repos 1-6
// for (var i = 0; i < 500; i++) {
// 	createTeam(i.toString(), "assignme");
// 	createRepo(i.toString(), "assignme");
// }

// get team data
// getTeamData("assignme");
// writeJSON();

//Assign repository to teams:
// for (var i = 0; i < 3; i++) {
// 	addRepoToTeam(i, i, "assignme");
// }

//Adding users to teams:
// for (var i = 0; i < 200; i++) {
// 	// addUserToTeam(usernames[i], i, "assignme");
// 	addUserToTeam('ssmirr', i, "assignme");
// 	console.log(getTeamId(i, "assignme"));
// }

//Update repo permissions:
// for (var i = 0; i < 6; i++) {
// 	updateTeamRepositoryPermissions(i, i, "assignme", "pull");
// }

//removing users from teams:
// for (var i = 0; i < 6; i++) {
// 	removeUserFromTeam(usernames[i], i, "assignme");
// }

//remove repository 1-6 from teams 1-6
// for (var i = 0; i < 6; i++) {
// 	removeTeamRepository(i, i, "assignme");
// }

//delete team 1-6
// for (var i = 0; i < 200; i++) {
// 		// deleteTeam(getTeamId(i.toString(), "assignme"));
// 		deleteTeam(getTeamId(i.toString(), "assignme"));
// }

//delete repo 1-6
// for (var i = 0; i < 20; i++) {
// 	deleteRepo(i.toString(), "assignme");
// }

// reading repository data and assigning the team names
// assignAll('ncsu_repository_data');




//**************************************************************
//**************************************************************
//*********************** Instructions *************************
//**************************************************************
//**************************************************************




var args = process.argv.slice(2);
var command = args[0];
var org = args[1];
var repositories = args[2]; //unityid:teamname
var permission = args[3]; //new permission level: push/pull for updating use access level

/** handling the entered command */
if (command === 'getdata') {
  if (typeof command !== 'undefined') {
    if (typeof org == 'undefined')
      console.error('Second commandline parameter not entered: Organization Name');
    else {
      console.log('Get data ... \n\n');
      getTeamData(org);
    }
  }
}

//fixing the json file after concatinating all pages.
else if (command === 'makejson') {
  console.log('Making the json file ...\n\n');
  writeJSON();
}

//assign the repositories
else if (command === 'assign') {
  if (typeof org == 'undefined')
    console.error('Commandline parameter not entered: Organization_Name Repository_List_File');
  else if (typeof repositories == 'undefined')
    console.error('Third Commandline parameter not entered: Repository_List_File');
  else {
    console.log('Assigning the repositories ...\n\n');
    assignAll(repositories, org);
  }
}

//changing user access to repositories
else if (command === 'access') {
  if (typeof org == 'undefined')
    console.error('Commandline parameter not entered: Organization_Name Repository_List_File Access_Level');
  else if (typeof repositories == 'undefined')
    console.error('Commandline parameter not entered: Repository_List_File Access_Level');
  else if (typeof permission == 'undefined')
    console.error('4th Commandline parameter not entered: Access_Level');
  else {
    console.log('Assigning the repositories ...\n\n');
    changeAccess(repositories, org, permission.toString());
  }
} else
  console.error('Invalid parameter entered! use: getdata, makejson, assign');



// Here are the steps:

// *** 1 *** assignme is the organization name
// node assignme.js getdata engr-csc216-spring2016

// *** 2 *** Makes correction to json files after merging them
// node assignme.js makejson

// *** 3 *** P1_repos.cvs is the team data (on each line username:teamname)
// node assignme.js assign engr-csc216-spring2016 P1_repos.csv

// *** To change use access on repositories run: changeAccess
// node assignme.js access engr-csc216-spring2016 P1_repos.csv pull


// Please MAKE SURE run each of the steps above on by one (comment out the one that you don't want to run)
// otherwise it might nto work correctly.
// ** This is the next thing I need to fix **

File: python/assignme.py
import sys, getopt
from os import system
from os.path import isfile
import sqlite3
import json as JSON
try:
	import requests
except ImportError:
	sys.exit('\033[1;31mError, Module requests is required\033[1;0m')

token = open('token', 'r').read()

# API_URL = 'https://api.github.com/'
API_URL = 'https://github.ncsu.edu/api/v3/'

headers = {
    'User-Agent': 'AssignMe',
    'content-type': 'application/json',
    'Authorization': token
}


def makeDb(csvPath, dbPath, tableName):
	if(isfile(dbPath)):
		print('\033[1;31m -> There already exists a database with the given dbPath, try again ...\033[1;0m\n')
		return
	else:
		print(' * Making the db file ...')
		print(' * Importing from csv ...')
		system('echo ".import ' + csvPath + ' ' + tableName + '" | sqlite3 ' + dbPath + ' -csv')

		conn = sqlite3.connect(dbPath)
		c = conn.cursor()
		c.execute('ALTER TABLE {tn} RENAME TO TEMPOLDTABLE;'.format(tn=tableName))
		c.execute('CREATE TABLE {tn} (unityId, teamName, teamId DEFAULT NULL, assigned DEFAULT NULL)'.format(tn=tableName))
		c.execute('INSERT INTO {tn} (unityId, teamName) SELECT unityId, teamName FROM TEMPOLDTABLE;'.format(tn=tableName))
		c.execute('DROP TABLE TEMPOLDTABLE;')
		conn.commit()
		conn.close()

# makeDb('fall2016.csv', 'fall2016.db', 'fall2016')

def getTeamId(teamName, dbPath, tableName):
	conn = sqlite3.connect(dbPath)
	c = conn.cursor()
	c.execute('select teamId from {tn} where teamName = "{teamName}";'.format(tn=tableName, teamName=teamName))
	teamId = c.fetchone()[0];
	conn.commit()
	conn.close()
	return teamId

def createRepo(repoName, orgName):
	json = {
		'name': repoName,
		'auto_init': 'true',
		'public': 'false',
		'gitignore_template': 'nanoc'
	}

	url = '{api}orgs/{orgName}/repos'.format(api=API_URL, orgName=orgName)

	r = requests.post(url, headers=headers, json=json)
	statusCode = int(r.headers['Status'].split(' ')[0])

	if statusCode >= 400:
		print 'Server error: ', statusCode
		return
	print 'Created repository {repoName} in {orgName}'.format(repoName=repoName, orgName=orgName)

# createRepo('test2', 'assignme')

def deleteRepo(repoName, orgName):
	url = '{api}teams/{id}'.format(api=API_URL, id=id)
	r = requests.delete(url, headers=headers)
	statusCode = int(r.headers['Status'].split(' ')[0])
	if statusCode >= 400:
		print 'Server error: ', statusCode
		return

	print 'Deleted the team with id = {id}'.format(id=id)


def deleteRepo(repoName, orgName):
	url = '{api}repos/{orgName}/{repoName}'.format(api=API_URL, orgName=orgName, repoName=repoName)

	r = requests.delete(url, headers=headers)
	statusCode = int(r.headers['Status'].split(' ')[0])

	if statusCode >= 400:
		print 'Server error: ', statusCode
		return
	print 'Deleted the repository {repoName} in {orgName}'.format(repoName=repoName, orgName=orgName)

# deleteRepo('test2', 'assignme')


def createTeam(teamName, orgName, permission, dbPath, tableName):
	json = {
		'name': teamName,
		'repo_names': ['{orgName}/{teamName}'.format(orgName=orgName, teamName=teamName)],
		'privacy': 'secret',
		'permission': permission
	}

	url = '{api}orgs/{orgName}/teams'.format(api=API_URL, orgName=orgName)

	r = requests.post(url, headers=headers, json=json)
	statusCode = int(r.headers['Status'].split(' ')[0])

	if statusCode >= 400:
		print 'Server error: ', statusCode
		return

	print 'Created the team {teamName} in {orgName}'.format(teamName=teamName, orgName=orgName)

	conn = sqlite3.connect(dbPath)
	c = conn.cursor()
	c.execute('UPDATE {tn} set teamId = ? where teamName = "{teamName}";'.format(tn=tableName, teamName=teamName), (JSON.loads(r.text)['id'],))
	conn.commit()
	conn.close()

# createTeam('test'+str(x), 'assignme', ['AssignMe/test1', 'AssignMe/test2'], 'push', 'fall2016.db')


def deleteTeam(teamName, dbPath, tableName):
	url = '{api}teams/{id}'.format(api=API_URL, id=getTeamId(teamName, dbPath, tableName))
	r = requests.delete(url, headers=headers)
	statusCode = int(r.headers['Status'].split(' ')[0])
	if statusCode >= 400:
		print 'Server error: ', statusCode
		return
	print 'Deleted the team with teamName = {teamName}'.format(teamName=teamName)
	conn = sqlite3.connect(dbPath)
	c = conn.cursor()
	c.execute('UPDATE {tn} set teamId = NULL where teamName = "{teamName}";'.format(tn=tableName, teamName=teamName))
	conn.commit()
	conn.close()


def addUserToTeam(userName, teamName, orgName, dbPath, tableName):
	url = '{api}teams/{id}/memberships/{userName}'.format(api=API_URL, id=getTeamId(teamName, dbPath, tableName), userName=userName)
	r = requests.put(url, headers=headers)
	statusCode = int(r.headers['Status'].split(' ')[0])
	if statusCode >= 400:
		print 'Server error: ', statusCode
		return
	print 'Added {userName} to {teamName}'.format(userName=userName, teamName=teamName)
	

def updateTeamRepoPermission(teamName, orgName, repoName, permission, dbPath, tableName):
	json = {
		'permission': permission
	}
	url = '{api}teams/{id}/repos/{orgName}/{repoName}'.format(api=API_URL, id=getTeamId(teamName, dbPath, tableName), orgName=orgName, repoName=repoName)
	r = requests.put(url, headers=headers, json=json)
	statusCode = int(r.headers['Status'].split(' ')[0])
	if statusCode >= 400:
		print 'Server error: ', statusCode
		return
	print 'Updated the team permission to {permission} where teamName = {teamName} '.format(teamName=teamName, permission=permission)


##################################################################
# Handling commandline args ######################################
##################################################################

# Read command line args
myopts, args = getopt.getopt(sys.argv[1:],"mud", ["dbPath=", "permission=", "tableName=", "import=", "orgName=", "assign"])

tableName = ''
dbPath = ''
permission = ''
importing = False
csv = ''
orgName = ''
makingTeams = False
updatingPermission = False
deletingTeams = False
assigningUsers = False

for o, a in myopts:
	# print '\nYES', o, a
	if o == '-m':
		makingTeams = True
	elif o == '-u':
		updatingPermission = True
	elif o == '-d':
		deletingTeams = True
	elif o == '--permission':
		permission = a
	elif o == '--tableName':
		tableName = a
	elif o == '--dbPath':
		dbPath = a
	elif o == '--import':
		importing = True
		csv = a
	elif o == '--orgName':
		orgName = a
	elif o == '--assign':
		assigningUsers = True
	else: 
		print("Usage: %s -makeTeamsAndRepos" % sys.argv[0])

# print tableName
# print dbPath
# print permission
# print makingTeams
# print updatingPermission
# print deletingTeams
# print csv
# print orgName
# print importing



if importing:
	makeDb(csv, dbPath, tableName)
	sys.exit(0)

elif makingTeams:
	conn = sqlite3.connect(dbPath)
	c = conn.cursor()
	c.execute('select teamName from {tn} where teamId is NULL'.format(tn=tableName))
	teams = c.fetchall()
	for team in teams:
		createRepo(team[0], orgName)
		createTeam(team[0], orgName, 'push', dbPath, tableName)
	conn.commit()
	conn.close()

elif updatingPermission:
	conn = sqlite3.connect(dbPath)
	c = conn.cursor()
	c.execute('select teamName from {tn} where teamId is not NULL'.format(tn=tableName))
	teams = c.fetchall()
	for team in teams:
		updateTeamRepoPermission(team[0], orgName, team[0], permission, dbPath, tableName)
	conn.commit()
	conn.close()

elif deletingTeams:
	conn = sqlite3.connect(dbPath)
	c = conn.cursor()
	c.execute('select teamName from {tn} where teamId is not NULL'.format(tn=tableName))
	teams = c.fetchall()
	for team in teams:
		deleteTeam(team[0], dbPath, tableName)
		deleteRepo(team[0], orgName)
	conn.commit()
	conn.close()

elif assigningUsers:
	conn = sqlite3.connect(dbPath)
	c = conn.cursor()
	c.execute('select unityId, teamName from {tn} where assigned is NULL'.format(tn=tableName))
	rows = c.fetchall()
	for row in rows:
		addUserToTeam(row[0], row[1], orgName, dbPath, tableName)
	conn.commit()
	conn.close()

# $ python assignme.py --import fall2016.csv --dbPath fall2016.db --tableName fall2016
# $ python assignme.py -m --dbPath fall2016.db --orgName assignme --tableName fall2016
# $ python assignme.py --assign --dbPath fall2016.db --orgName assignme --tableName fall2016
# $ python assignme.py -u --dbPath fall2016.db --orgName assignme --tableName fall2016 --permission pull
# $ python assignme.py -d --dbPath fall2016.db --orgName assignme --tableName fall2016
